egui/src/widgets/text_edit/validation.rs

```diff
@@ -0,0 +1,204 @@
+                        cmp::Ordering::Less => Action::Insert(filter),
+                        // Guaranteed to exceed the limit so we overwrite.
+                        cmp::Ordering::Greater => Action::Overwrite(Cow::Borrowed(min_str)),
+                        // Slow case where we need to actually compare with the limit.
+                        cmp::Ordering::Equal => {
+                            // Note that we don't parse the strings because its too slow.
+                            let mut iter = buffer.chars();
+                            let front = min_str[..buffer.len()].chars();
+                            for (current, min) in (&mut iter).zip(front) {
+                                if current < min {
+                                    return Action::Overwrite(Cow::Borrowed(min_str));
+                                }
+                            }
+
+                            let back = min_str[buffer.len()..].chars();
+                            for (current, min) in iter.zip(back) {
+                                if current > min {
+                                    return Action::Overwrite(Cow::Borrowed(min_str));
+                                }
+                            }
+
+                            Action::Insert(filter)
+                        }
+                    }
+                } else {
+                    match new_size.cmp(&max_str_len) {
+                        // Guaranteed to be below the limit so we insert.
+                        cmp::Ordering::Less => Action::Insert(filter),
+                        // Guaranteed to exceed the limit so we overwrite.
+                        cmp::Ordering::Greater => Action::Overwrite(Cow::Borrowed(max_str)),
+                        // Slow case where we need to actually compare with the limit.
+                        cmp::Ordering::Equal => {
+                            // Note that we don't parse the strings because its too slow.
+                            let mut iter = buffer.chars();
+                            let front = max_str[..buffer.len()].chars();
+                            for (current, max) in (&mut iter).zip(front) {
+                                if current > max {
+                                    return Action::Overwrite(Cow::Borrowed(max_str));
+                                }
+                            }
+
+                            let back = max_str[buffer.len()..].chars();
+                            for (current, max) in iter.zip(back) {
+                                if current > max {
+                                    return Action::Overwrite(Cow::Borrowed(max_str));
+                                }
+                            }
+
+                            Action::Insert(filter)
+                        }
+                    }
+                }
+            }
+        }
+    };
+}
+
+impl_validate_input_signed!(i8, "127", "-128");
```

Comment:
nitpicky, but `min` should come before `max`

Reply:
agreed